plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,361 @@
+			return getArithmeticMultiaryType((MultiplyExpression) expression);
+		}
+		if (expression instanceof ArrayAccessExpression) {
+			Expression tmpExpression = ((ArrayAccessExpression) expression).getOperand();
+			Type declarationType = expressionUtil.getDeclaration(tmpExpression).getType();
+			if (declarationType instanceof ArrayTypeDefinition) {
+				return ((ArrayTypeDefinition) declarationType).getElementType();
+			}
+		}
+		if (expression instanceof FunctionAccessExpression) {
+			Type declarationType = expressionUtil.getDeclaration(expression).getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof RecordAccessExpression) {
+			RecordAccessExpression recordAccessExpression = (RecordAccessExpression) expression;
+			return getType(recordAccessExpression.getFieldReference());
+		}
+		if (expression instanceof SelectExpression) {
+			Expression tmpExpression = ((SelectExpression) expression).getOperand();
+			Type declarationType = expressionUtil.getDeclaration(tmpExpression).getType();
+			TypeDefinition typeDefinition = removeTypeReferences(declarationType);
+			if (typeDefinition instanceof ArrayTypeDefinition) {
+				ArrayTypeDefinition arrayTypeDefinition = (ArrayTypeDefinition) typeDefinition;
+				return removeTypeReferences(arrayTypeDefinition.getElementType());
+			}
+			else if (typeDefinition instanceof IntegerRangeTypeDefinition) {
+				return factory.createIntegerTypeDefinition();
+			}
+			else if (typeDefinition instanceof EnumerationTypeDefinition) {
+				return typeDefinition;
+			}
+			else {
+				throw new IllegalArgumentException("The type of the operand of the select expression is not an enumerable type: " + expressionUtil.getDeclaration(expression));
+			}
+		}
+		if (expression == null) {
+			return factory.createVoidTypeDefinition();
+		}
+		throw new IllegalArgumentException("Unknown type!");
+	}
+	
+	// Equals, TypeReference
+	
+	public TypeDefinition removeTypeReferences(Type type) {
+		Type clone = ecoreUtil.clone(type);
+		for (TypeReference reference : ecoreUtil.getAllContentsOfType(clone, TypeReference.class)) {
+			TypeDefinition typeDefinition = ExpressionModelDerivedFeatures.getTypeDefinition(reference);
+			ecoreUtil.replace(reference, typeDefinition);
+		}
+		return ExpressionModelDerivedFeatures.getTypeDefinition(clone);
+	}
+	
+	public TypeDefinition getTypeDefinition(Expression expression) {
+	    return removeTypeReferences(getType(expression));
+	}
+	
+	public boolean equals(Expression expressionOne, Expression expressionTwo) {
+		Type typeOne = getType(expressionOne);
+		Type typeTwo = getType(expressionTwo);
+		return equals(typeOne, typeTwo);
+	}
+	
+	public boolean equals(Type typeOne, Expression expression) {
+		Type typeTwo = getType(expression);
+		return equals(typeOne, typeTwo);
+	}
+	
+	public boolean equals(Type typeOne, Type typeTwo) {
+		if (typeOne instanceof TypeReference) {
```

Comment:
`removeTypeReferences(...);` should be used regardless (remove the `if` statements).